File: ios/PlayKit/PlayKit/PlayKitQosModuleWrapper.m
    //
    //  PlayKitQosModuleWrapper.m
    //  PlayKit
    //
    //  Created by Lamine Ndiaye on 21/06/2019.
    //  Copyright © 2019 Kaltura. All rights reserved.
    //

#import "PlayKitQosModuleWrapper.h"
#import <AVFoundation/AVFoundation.h>

@interface PlayKitQosModuleWrapper()
@property (strong, nonatomic) id<Player> player;
@end

@implementation PlayKitQosModuleWrapper
+(instancetype)playKitQoSModule:(id<Player>) player {
    PlayKitQosModuleWrapper *playkitQosModule = [[PlayKitQosModuleWrapper alloc] init];
    playkitQosModule.player = player;
    return playkitQosModule;
}

- (instancetype)init {
    if (self = [super init]) {
        self.dnaQoSModule = [DNAQos moduleWithType:QOSModuleTypeCustom];
        [self registerObservers];
    }
    return self;
}

-(void)registerObservers {
    __weak typeof(self) weakSelf = self;
    [self.player addObserver: self event: PlayerEvent.stateChanged block:^(PKEvent* event)  {
        typeof(self) strongSelf = weakSelf;
        switch (event.newState) {
            case PlayerStateIdle:
                [strongSelf.dnaQoSModule playerStateDidChange: PlaybackStateIdle];
            case PlayerStateReady:
                if (strongSelf.player.isPlaying) {
                    [strongSelf.dnaQoSModule playerStateDidChange: PlaybackStatePlaying];
                } else {
                    [strongSelf.dnaQoSModule playerStateDidChange: PlaybackStatePaused];
                }
            case PlayerStateBuffering:
                [strongSelf.dnaQoSModule playerStateDidChange: PlaybackStateBuffering];
            case PlayerStateError:
                [strongSelf.dnaQoSModule playbackErrorOccurred];
            default:
                break;
        }
    }];
    
    [self.player addObserver: self event: PlayerEvent.videoTrackChanged block:^(PKEvent* event) {
        typeof(self) strongSelf = weakSelf;
        [strongSelf.dnaQoSModule trackSwitchOccurred];
    }];
    
    [self.player addObserver: self event: PlayerEvent.seeked block:^(PKEvent* event) {
        typeof(self) strongSelf = weakSelf;
        [strongSelf.dnaQoSModule playerStateDidChange: PlaybackStateSeeking];
    }];
    
    [self.player addObserver: self event: PlayerEvent.playbackStalled block:^(PKEvent* event) {
        typeof(self) strongSelf = weakSelf;
        [strongSelf.dnaQoSModule playbackErrorOccurred];
    }];
    
    [self.player addObserver: self event: PlayerEvent.ended block:^(PKEvent* event) {
        typeof(self) strongSelf = weakSelf;
        [strongSelf.dnaQoSModule playerStateDidChange: PlaybackStateEnded];
    }];
    
    [self.player addObserver: self event: PlayerEvent.error block:^(PKEvent* event) {
        typeof(self) strongSelf = weakSelf;
        [strongSelf.dnaQoSModule playbackErrorOccurred];
    }];
}

- (void)dealloc
{
    [self.player removeObserver: self events: @[PlayerEvent.stateChanged,
                                                PlayerEvent.videoTrackChanged,
                                                PlayerEvent.seeked,
                                                PlayerEvent.playbackStalled,
                                                PlayerEvent.ended,
                                                PlayerEvent.error]];
}
@end

File: ios/PlayKit/PlayKit-Swift/PlayKitQosModuleWrapper.swift
//
//  PlayKitQoSModule.swift
//  PlayKitOVPStarter
//
//  Created by Jean-Baptiste Louazel on 13/06/2019.
//  Copyright © 2019 Kaltura. All rights reserved.
//

import PlayKit
import StreamrootSDK

// Streamroot QoS module
// Binds player' events to raise precise data to DNAClient
class PlayKitQosModuleWrapper {
  let dnaQoSModule: QosModule = DNAQos.module(type: .custom)
    let player: Player

    public init(_ player: Player) {
        self.player = player
        player.addObserver(self, event: PlayerEvent.stateChanged, block: { event in
            switch (event.newState) {
            case .idle:
                self.dnaQoSModule.playerStateDidChange(PlaybackState.idle)
            case .ready:
                player.isPlaying
                    ? self.dnaQoSModule.playerStateDidChange(PlaybackState.playing)
                    : self.dnaQoSModule.playerStateDidChange(PlaybackState.paused)
            case .buffering:
                self.dnaQoSModule.playerStateDidChange(PlaybackState.buffering)
            case .ended:
                self.dnaQoSModule.playerStateDidChange(PlaybackState.ended)
            case .error:
                self.dnaQoSModule.playbackErrorOccurred()
            default:
                break
            }
        })

        player.addObserver(self, event: PlayerEvent.videoTrackChanged, block: { _ in self.dnaQoSModule.trackSwitchOccurred() })

        player.addObserver(self, event: PlayerEvent.seeked, block: { _ in self.dnaQoSModule.playerStateDidChange(PlaybackState.seeking) })
        
        player.addObserver(self, event: PlayerEvent.playbackStalled, block: { _ in self.dnaQoSModule.playbackErrorOccurred() })
        
        player.addObserver(self, event: PlayerEvent.ended, block: { _ in self.dnaQoSModule.playerStateDidChange(PlaybackState.ended) })
        
        player.addObserver(self, event: PlayerEvent.error, block: { _ in self.dnaQoSModule.playbackErrorOccurred() })
    }
    
    deinit {
        player.removeObserver(self, events: [PlayerEvent.stateChanged,
                                             PlayerEvent.videoTrackChanged,
                                             PlayerEvent.seeked,
                                             PlayerEvent.playbackStalled,
                                             PlayerEvent.ended,
                                             PlayerEvent.error])
    }
    
}
